File: RxJavaRetrofit/app/src/main/java/com/woowrale/rxjavaretrofit/ui/main/MainPresenter.java
package com.woowrale.rxjavaretrofit.ui.main;

import com.woowrale.rxjavaretrofit.data.model.Crypto;
import com.woowrale.rxjavaretrofit.data.repository.ApiService;
import com.woowrale.rxjavaretrofit.ui.base.BasePresenter;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;

import java.util.List;

public class MainPresenter extends BasePresenter<MainView> {

    private CompositeDisposable compositeDisposable;

    public void onSearchCryptoCurrency(){

        compositeDisposable = new CompositeDisposable();
        compositeDisposable.add(ApiService.services().getCoinData("btc")
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(this::handleResults));

        /*
        Observable<List<Crypto.Market>> btcObservable = ApiRestServices.services().getCoinData("btc")
                .map(result -> Observable.fromIterable(result.ticker.markets))
                .flatMap(x -> x).filter(y -> {
                    y.coinName = "btc";
                    return true;
                }).toList().toObservable();

        Observable<List<Crypto.Market>> ethObservable =  ApiRestServices.services().getCoinData("eth")
                .map(result -> Observable.fromIterable(result.ticker.markets))
                .flatMap(x -> x).filter(y -> {
                    y.coinName = "eth";
                    return true;
                }).toList().toObservable();

        Observable.merge(btcObservable, ethObservable)
                .subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::handleResults, this::handleError);
         */
    }


    private void handleResults(Crypto crypto) {
        if (crypto != null) {
            if(crypto.ticker.markets != null){
                getView().showListView(crypto.ticker.markets);
            }
        } else {
            getView().showMessageResult();
        }
    }

    private void handleError(Throwable t) {
        getView().showMessageError();
    }

}

File: RxKotlinRetrofit/app/src/main/java/com/woowrale/rxkotlinretrofit/ui/main/MainPresenter.kt
package com.woowrale.rxkotlinretrofit.ui.main

import android.annotation.SuppressLint
import android.util.Log
import com.woowrale.rxkotlinretrofit.data.model.Market
import com.woowrale.rxkotlinretrofit.data.repository.ApiServices
import com.woowrale.rxkotlinretrofit.data.repository.RestClient
import com.woowrale.rxkotlinretrofit.ui.base.BasePresenter
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers

class MainPresenter : BasePresenter<MainView>() {

    val TAG_NAME: String = this.javaClass.canonicalName

    @SuppressLint("CheckResult")
    fun onCallEndPoint() {
        val cryptocurrencyService = RestClient.getRetrofit().create(ApiServices::class.java)

        //Single call
        /*Observable<Crypto> cryptoObservable = cryptocurrencyService.getCoinData("btc");
        cryptoObservable.subscribeOn(Schedulers.newThread()).observeOn(AndroidSchedulers.mainThread()).map(result -> result.ticker).subscribe(this::handleResults, this::handleError);*/
        val btcObservable = cryptocurrencyService.getCoinData("btc")
            .map { result -> Observable.fromIterable(result.ticker?.markets) }
            .flatMap({ x -> x }).filter({ y ->
                y.coinName = "btc"
                true
            }).toList().toObservable()
        /*
        val ethObservable = cryptocurrencyService.getCoinData("eth")
            .map { result -> Observable.fromIterable(result.ticker?.markets) }
            .flatMap({ x -> x }).filter({ y ->
                y.coinName = "eth"
                true
            }).toList().toObservable()
        */
        Observable.merge<List<Market>>(btcObservable, btcObservable)
            .subscribeOn(Schedulers.computation())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe({ this.handleResults(it) }, { this.handleError(it) })
    }

    private fun handleResults(marketList: List<Market>?) {
        if (marketList != null && marketList.size != 0) {
            getView()?.setRecyclerViewAdapterData(marketList)
        } else {
            getView()?.showMessageResult()
        }
    }

    private fun handleError(t: Throwable) {
        Log.e(TAG_NAME, t.message)
        getView()?.showMessageError()
    }
}